### PR TITLE
Net perf minimal

### PR DESCRIPTION
**Pull request by dustinblack**
## Changes introduced with this PR

Initial draft for net-perf workflow yaml.

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: network-performance/network-performance.yaml [hunk 1]
```diff
@@ -184,57 +184,63 @@ steps:
...
+          #TODO -- How to tie a client to a server?
+          target: 
   #new plugin needed
   es_translator:
+    id: es_translator
     needs:
-      - steps.metadata_collector.output["success"]
-      - steps.metrics_collector.output["success"]
-      - steps.run_uperf_client.output["success"]
+      - $.steps.metadata_collector.success.output
```
[the remark this answers is not in the document]
Comment (dustinblack) [reply network-performance/network-performance.yaml:237]: I'm not following what you mean here

File: network-performance/network-performance-workflow.yaml [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            imagePullPolicy: Always
+    input:
+      port: 20000
+      name: "netperf"
+      groups:
+        - nthreads: 1
+          transactions:
+            - iterations: 1
+              flowops:
+                - type: "accept"
```
Comment (jaredoconnell) [inline network-performance/network-performance-workflow.yaml:124]: I don't know if it currently works with `accept`. `connect` was the one that worked in the demo.

File: network-performance/README.md [hunk 3]
```diff
@@ -1,202 +1,196 @@
...
-    cluster_name: str
-    prom_token: str
-```
-```mermaid
-graph TD
-    global_params[global params] --> metrics_plugin([SUT metrics collection])
-    global_params --> metadata_plugin([SUT metadata collection])
-    global_params --> pod_config(k8s pod/job config)
-```
+input:
```
Comment (portante) [inline network-performance/README.md:15]: Do we have a way to keep this `yaml` updated with the external file?  Perhaps have a way to re-generate the `README.md` file from a template that includes the external file?

File: network-performance/README.md [hunk 4]
```diff
@@ -1,202 +1,196 @@
...
-
-    pod_config(k8s pod/job config) --> scheduling
-    scheduling(k8s scheduling) --> uperf_servers
-    scheduling --> uperf_clients
-
-    uperf_servers([uperf servers])
-    uperf_clients([uperf clients]) --> es
-    es([elasticsearch])
-```
+steps:
```
Comment (portante) [inline network-performance/README.md:89]: Same comment as above ... this `yaml` can grow stale on the vine very quickly.

File: network-performance/network-performance-workflow.yaml [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        uuid: !expr $.steps.uuidgen.outputs.success.uuid
+        run_id: !expr $.input.run_id
+        uperf: !expr $.steps.uperf_client.outputs.success
+        metadata: !expr $.steps.metadata.outputs.success
+output:
+  uuid: !expr $.steps.uuidgen.outputs.success.uuid
+  run_id: !expr $.input.run_id
+  uperf: !expr $.steps.uperf_client.outputs.success
+  metadata: !expr $.steps.metadata.outputs.success
+
```
Comment (portante) [inline network-performance/network-performance-workflow.yaml:164]: Looks like a trailing whitespace issue here.

**Comment by dustinblack** [comment]
Initially the pseudo-code was in the README because we didn't yet know what a real workflow should look like. I really don't think it's necessary here now, nor is the detailed documentation. The docs should move somewhere else with examples, and this should be whittled down to just the workflow fundamental files and a descriptive README. I'll work on that.

File: network-streaming-performance-k8s/README.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
+# Network Streaming Performance Workflow for Kubernetes
+
+***NOTE: This is a work-in-progress workflow.***
+
+## Workflow Description
+
+This example workflow runs a network streaming bi-directional load using the
+[uperf](https://github.com/uperf/uperf) benchmark utility. The workload will
```
Comment (portante) [inline network-streaming-performance-k8s/README.md:8]: Consider using `workflow` instead of `workload` here.

Comment (dustinblack) [reply network-streaming-performance-k8s/README.md:8]: "Workload" is intentional here. The "workflow" runs plugins both locally and on the k8s cluster, so "workload" is used to differentiate the plugins running an actual load, which will be scheduled on the k8s cluster.

Comment (portante) [reply network-streaming-performance-k8s/README.md:8]: I still find it confusing. But perhaps by working in the explanation you gave above into the text we could avoid that.

File: network-streaming-performance-k8s/README.md [hunk 7]
```diff
@@ -4,11 +4,11 @@
 
 ## Workflow Description
 
-This example workflow runs a network streaming bi-directional load using the
+This example workflow runs a network streaming one-directional workload using the
 [uperf](https://github.com/uperf/uperf) benchmark utility. The workload will
 run on a Kubernetes cluster using a provided `kubeconfig` object and will use a kubernetes service for communication between the uperf server and client pods.
 
-In addition to the uperf load, the workflow generates a UUID, which can be used as a unique key for the generated data, and runs a metadata collection using Ansible [gather facts](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/gather_facts_module.html).
+In addition to the uperf workload, the workflow generates a UUID, which can be used as a unique key for the generated data, collects system metrics with [Performance Co-pilot](https://pcp.io/), and collects system metadata using Ansible [gather facts](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/gather_facts_module.html).
```
Comment (portante) [inline network-streaming-performance-k8s/README.md:11]: This system metrics and system metadata collection is a re-implementation of what other systems already do today.  I realize this is new, but we run the risk of proliferating re-implementations to work with Arcaflow if we are not careful.

**Review by portante — CHANGES_REQUESTED** [review]
The `README.md` is still confusing to me with "workflow" and "workload".

File: network-streaming-performance-k8s/README.md (hunk 6 above)
Comment (dustinblack) [reply network-streaming-performance-k8s/README.md:8]: I updated the language a bit, so I'll resolve this one for now.

File: network-streaming-performance-k8s/README.md (hunk 7 above)
Comment (dustinblack) [reply network-streaming-performance-k8s/README.md:11]: Truly the plugins that are in place today for metadata and metrics are rudimentary proof-of-concept implementations. Replacing them with plugins that better leverage existing implementations will likely be desired.

That being said, we should effectively be agnostic about the functions of individual plugins, and there isn't any reason that multiple approaches to what may look like the same challenge can't coexist.

The PR here is for a workflow to serve as a functional example, and so it is somewhat independent from the plugins it uses. I don't think that concerns about plugin implementations should necessarily block this workflow example PR.

Comment (jaredoconnell) [reply network-streaming-performance-k8s/README.md:11]: Yeah. We may package https://github.com/cloud-bulldozer/metadata-collector in a redhat-specific repository.

**Review by portante — APPROVED** [review]

**Pull request merged**